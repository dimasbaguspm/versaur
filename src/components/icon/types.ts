/**
/**
 * IconProps defines the props for the Icon component
 * @property color - Color variant based on Versaur color system
 * @property size - Size of the icon (xs, sm, md, lg, xl)
 * @property children - The icon element (usually from lucide-react)
 * Extends HTMLAttributes<HTMLSpanElement> for native span props
 */
/**
 * IconProps defines the props for the Icon component
 * @property as - Icon component to render (e.g., from lucide-react)
 * @property color - Color variant based on Versaur color system
 * @property size - Size of the icon (xs, sm, md, lg, xl)
 * Extends SVG props for native SVG attributes
 */
export type IconProps = Omit<React.SVGProps<SVGSVGElement>, 'children'> & {
  /**
   * Icon component to render (e.g., from lucide-react)
   */
  as: React.ComponentType<{ className?: string }>
  /**
   * Color variant
   */
  color?:
    | 'primary'
    | 'secondary'
    | 'tertiary'
    | 'ghost'
    | 'neutral'
    | 'success'
    | 'info'
    | 'warning'
    | 'danger'
  /**
   * Size of the icon
   */
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl'
}
