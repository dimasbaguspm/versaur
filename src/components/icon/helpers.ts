import { cva } from 'class-variance-authority'

export const iconVariants = cva('inline-flex items-center justify-center', {
  variants: {
    size: {
      xs: 'w-3 h-3',
      sm: 'w-4 h-4',
      md: 'w-5 h-5',
      lg: 'w-6 h-6',
      xl: 'w-8 h-8',
    },
    color: {
      primary: 'text-primary',
      secondary: 'text-secondary',
      tertiary: 'text-tertiary',
      ghost: 'text-ghost',
      neutral: 'text-neutral',
      success: 'text-success',
      info: 'text-info',
      warning: 'text-warning',
      danger: 'text-danger',
    },
  },
  defaultVariants: {
    size: 'md',
    color: 'primary',
  },
})
