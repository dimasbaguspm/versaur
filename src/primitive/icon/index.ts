export { Icon } from './icon'
export type * from './types'
