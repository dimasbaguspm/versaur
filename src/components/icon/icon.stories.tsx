/**
 * Icon stories for Versaur UI
 * Demonstrates color and size variations using Lucide icons
 * Shows how to use Icon with Lucide icons and native span props
 */
import { Icon } from './icon'
import { Circle, Star } from 'lucide-react'

export default {
  title: 'Components/Icon',
  component: Icon,
  tags: ['autodocs'],
}

export const Basic = {
  render: () => (
    <Icon color='primary' size='md' title='Primary Icon'>
      <Circle />
    </Icon>
  ),
  name: 'Basic usage',
  parameters: {
    docs: {
      description: {
        story:
          'A basic Icon wrapping a Lucide icon with primary color and medium size.',
      },
    },
  },
}

export const AllSizes = {
  render: () => (
    <div className='flex gap-2 items-center'>
      <Icon size='xs'>
        <Circle />
      </Icon>
      <Icon size='sm'>
        <Circle />
      </Icon>
      <Icon size='md'>
        <Circle />
      </Icon>
      <Icon size='lg'>
        <Circle />
      </Icon>
      <Icon size='xl'>
        <Circle />
      </Icon>
    </div>
  ),
  name: 'All sizes',
  parameters: {
    docs: {
      description: {
        story: 'Showcases all available icon sizes.',
      },
    },
  },
}

export const AllColors = {
  render: () => (
    <div className='flex gap-2 items-center'>
      <Icon color='primary'>
        <Star />
      </Icon>
      <Icon color='secondary'>
        <Star />
      </Icon>
      <Icon color='tertiary'>
        <Star />
      </Icon>
      <Icon color='ghost'>
        <Star />
      </Icon>
      <Icon color='neutral'>
        <Star />
      </Icon>
      <Icon color='success'>
        <Star />
      </Icon>
      <Icon color='info'>
        <Star />
      </Icon>
      <Icon color='warning'>
        <Star />
      </Icon>
      <Icon color='danger'>
        <Star />
      </Icon>
    </div>
  ),
  name: 'All colors',
  parameters: {
    docs: {
      description: {
        story: 'Showcases all available color variants.',
      },
    },
  },
}
