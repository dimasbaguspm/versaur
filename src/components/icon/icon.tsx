import React from 'react'
import type { IconProps } from './types'
import { iconVariants } from './helpers'

/**
 * Icon component for Versaur UI
 * Wraps Lucide icons and enforces consistent size and color
 * @example <Icon color="primary" size="md"><LucideIcon /></Icon>
 */
export const Icon: React.FC<IconProps> = ({
  color = 'primary',
  size = 'md',
  children,
  ...rest
}) => {
  return (
    <span
      className={iconVariants({ color, size, className: rest.className })}
      aria-hidden={'true'}
      {...rest}
    >
      {children}
    </span>
  )
}
