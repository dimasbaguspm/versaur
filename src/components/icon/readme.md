# Icon Component


Versaur UI Icon renders a single icon component (e.g., from lucide-react) with consistent size and color according to the Versaur design system. It enforces API consistency and removes the need for children.

## Usage

```tsx
import { Icon } from '@/components/icon'
import { Circle } from 'lucide-react'

<Icon as={Circle} color="primary" size="md" />
```

## Props
- `as`: **Required.** The icon component to render (e.g., from `lucide-react`).
- `color`: Versaur color variant (`primary`, `secondary`, `tertiary`, `ghost`, `neutral`, `success`, `info`, `warning`, `danger`). Default: `primary`.
- `size`: Icon size (`xs`, `sm`, `md`, `lg`, `xl`). Default: `md`.
- Accepts all native `<svg>` props (e.g., `id`, `aria-label`, `title`).

## Examples

- **Basic:**
  ```tsx
  <Icon as={Circle} color="primary" size="md" />
  ```
- **All Sizes:**
  ```tsx
  <Icon as={Circle} size="xs" />
  <Icon as={Circle} size="sm" />
  <Icon as={Circle} size="md" />
  <Icon as={Circle} size="lg" />
  <Icon as={Circle} size="xl" />
  ```
- **All Colors:**
  ```tsx
  <Icon as={Circle} color="danger" />
  <Icon as={Circle} color="info" />
  // ...other colors
  ```
- **Native svg props:**
  ```tsx
  <Icon as={Circle} id="custom-id" aria-label="icon-label" />
  ```

## Accessibility
- Icons are rendered with `aria-hidden="true"` by default.
- Pass `aria-label` or `role` for accessible icons as needed.

## Testing
- Snapshot and accessibility tests are provided in `__tests__/icon.test.tsx`.

## Storybook
- See `icon.stories.tsx` for interactive examples and documentation.
