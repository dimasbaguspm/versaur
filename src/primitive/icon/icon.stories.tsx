/**
 * Icon stories for Versaur UI
 * Demonstrates color and size variations using Lucide icons
 * Shows how to use Icon with Lucide icons and native span props
 */
import { Icon } from './icon'
import { Circle, Star } from 'lucide-react'

export default {
  title: 'Primitive/Icon',
  component: Icon,
  tags: ['autodocs'],
}

export const Basic = {
  render: () => (
    <Icon as={Circle} color='primary' size='md' aria-label='Primary Icon' />
  ),
  name: 'Basic usage',
  parameters: {
    docs: {
      description: {
        story:
          'A basic Icon rendering a Lucide icon with primary color and medium size.',
      },
    },
  },
}

export const AllSizes = {
  render: () => (
    <div className='flex gap-2 items-center'>
      <Icon size='xs' as={Circle} />
      <Icon size='sm' as={Circle} />
      <Icon size='md' as={Circle} />
      <Icon size='lg' as={Circle} />
      <Icon size='xl' as={Circle} />
    </div>
  ),
  name: 'All sizes',
  parameters: {
    docs: {
      description: {
        story: 'Showcases all available icon sizes.',
      },
    },
  },
}

export const AllColors = {
  render: () => (
    <div className='flex gap-2 items-center'>
      <Icon as={Star} color='primary' />
      <Icon as={Star} color='secondary' />
      <Icon as={Star} color='tertiary' />
      <Icon as={Star} color='ghost' />
      <Icon as={Star} color='neutral' />
      <Icon as={Star} color='success' />
      <Icon as={Star} color='info' />
      <Icon as={Star} color='warning' />
      <Icon as={Star} color='danger' />
    </div>
  ),
  name: 'All colors',
  parameters: {
    docs: {
      description: {
        story: 'Showcases all available color variants.',
      },
    },
  },
}
