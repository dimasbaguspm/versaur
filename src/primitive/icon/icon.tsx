import React from 'react'
import type { IconProps } from './types'
import { iconVariants } from './helpers'

/**
 * Icon component for Versaur UI
 * Renders a single icon component (e.g., from lucide-react) with enforced size and color
 * @example <Icon as={Circle} color="primary" size="md" />
 */
export const Icon = React.forwardRef<SVGSVGElement, IconProps>(function Icon(
  { as: IconComponent, color = 'primary', size = 'md', className, ...rest },
  ref
) {
  return (
    <IconComponent
      // @ts-expect-error - Lucide icons don't have className prop, but we use it for styling
      ref={ref}
      className={iconVariants({ color, size, className })}
      {...rest}
    />
  )
})
