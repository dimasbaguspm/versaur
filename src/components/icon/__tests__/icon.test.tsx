/**
 * Unit tests for Icon component
 * Uses Vitest and React Testing Library
 */
import { describe, it, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { Icon } from '../icon'
import { Circle } from 'lucide-react'

// Basic render test
it('renders Icon with default props', () => {
  const { asFragment } = render(
    <Icon>
      <Circle />
    </Icon>
  )
  expect(asFragment()).toMatchSnapshot()
})

describe('Icon props', () => {
  it('applies color and size classes', () => {
    render(
      <Icon color='danger' size='lg' data-testid='icon'>
        <Circle />
      </Icon>
    )
    const icon = screen.getByTestId('icon')
    expect(icon.className).toContain('text-danger')
    expect(icon.className).toContain('w-6')
    expect(icon.className).toContain('h-6')
  })

  it('passes native span props', () => {
    render(
      <Icon id='custom-id' aria-label='icon-label'>
        <Circle />
      </Icon>
    )
    const icon = screen.getByLabelText('icon-label')
    expect(icon).toHaveAttribute('id', 'custom-id')
  })
})
