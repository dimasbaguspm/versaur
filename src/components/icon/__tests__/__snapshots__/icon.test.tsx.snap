// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders Icon with default props 1`] = `
<DocumentFragment>
  <span
    aria-hidden="true"
    class="inline-flex items-center justify-center w-5 h-5 text-primary"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-circle"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
    </svg>
  </span>
</DocumentFragment>
`;
